src/core/messages-grant-authorization.ts

```diff
@@ -0,0 +1,132 @@
+      // if the message is a Records interface message, get the RecordsWrite message associated with the record
+      const recordsMessage = messageToGet as RecordsWriteMessage | RecordsDeleteMessage;
+      const recordsWriteMessage = Records.isRecordsWrite(recordsMessage) ? recordsMessage :
+        await MessagesGrantAuthorization.getRecordsWriteMessageToAuthorize(tenant, recordsMessage, messageStore);
+
+      if (recordsWriteMessage.descriptor.protocol === incomingScope.protocol) {
+        // the record protocol matches the incoming scope protocol
+        return;
+      }
+
+      // we check if the protocol is the internal PermissionsProtocol for further validation
+      if (recordsWriteMessage.descriptor.protocol === PermissionsProtocol.uri) {
+        // get the permission scope from the permission message
+        // if the permission message is a revocation, the scope is fetched from the grant that is being revoked
+        let permissionScope!: PermissionScope;
+        if (recordsWriteMessage.descriptor.protocolPath === PermissionsProtocol.revocationPath) {
+          const grant = await PermissionsProtocol.fetchGrant(tenant, messageStore, recordsWriteMessage.descriptor.parentId!);
+          permissionScope = grant.scope;
+          return;
+        } else if (recordsWriteMessage.descriptor.protocolPath === PermissionsProtocol.grantPath) {
+          const grant = await PermissionGrant.parse(recordsWriteMessage as DataEncodedRecordsWriteMessage);
+          permissionScope = grant.scope;
+        } else if (recordsWriteMessage.descriptor.protocolPath === PermissionsProtocol.requestPath) {
+          const request = await PermissionRequest.parse(recordsWriteMessage as DataEncodedRecordsWriteMessage);
+          permissionScope = request.scope;
+        }
+
+        if (PermissionsProtocol.hasProtocolScope(permissionScope) && permissionScope.protocol === incomingScope.protocol) {
+          // the permissions record scoped protocol matches the incoming scope protocol
+          return;
+        }
+      }
+    } else if (messageToGet.descriptor.interface === DwnInterfaceName.Protocols) {
+      // if the message is a protocol message, it must be a `ProtocolConfigure` message
+      const protocolsConfigureMessage = messageToGet as ProtocolsConfigureMessage;
+      const configureProtocol = protocolsConfigureMessage.descriptor.definition.protocol;
+      if (configureProtocol === incomingScope.protocol) {
+        // the configured protocol matches the incoming scope protocol
+        return;
+      }
+    }
+
+    throw new DwnError(DwnErrorCode.MessagesGetVerifyScopeFailed, 'record message failed scope authorization');
+  }
+
+  /**
+   * Get the RecordsWriteMessage associated with the given RecordsDeleteMessage in order to authorize access.
+   */
+  private static async getRecordsWriteMessageToAuthorize(
+    tenant: string,
+    message: RecordsDeleteMessage,
```

Comment:
Consider a slight change in method signature and name, which would be more generic, readable, and potentially reusable.

```suggestion
  private static async getNewestRecordsWrite(
    tenant: string,
    recordId: string,
```

Optional extra mile, with the above change the method become generic thus can be moved else for reuse, though looks like we have a few variants of this but not quite the same.

Reply:
🔥 I moved it over to the `RecordsWrite` interface class, reads better and may be more useful elsewhere.